packages/cli/src/commands/build-cache/cache.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs-extra';
+import { resolve as resolvePath, relative as relativePath } from 'path';
+import { runPlain, runCheck } from '../../helpers/run';
+import { Options } from './options';
+import { extractArchive, createArchive } from './archive';
+
+const INFO_FILE = '.backstage-build-cache';
+
+export type CacheQueryResult = {
+  hit: boolean;
+  copy?: (outputDir: string) => Promise<void>;
+  archive?: (outputDir: string, maxEntries: number) => Promise<void>;
```

Comment:
i'd like some comments on these, it's not immediately apparent what the different possible states are here (or maybe we could even make several distinct result types)

Reply:
yeah had some comments explaining the previous implementation but got rid of them for this ._. can explain a bit more, especially if we think this is sane